File: src/light.go
package src

import (
	"fmt"
	"unsafe"

	rl "github.com/gen2brain/raylib-go/raylib"
)

type LightType int32

var fstr = fmt.Sprintf

const (
	LightTypeDirectional LightType = iota
	LightTypePoint
)

type Light struct {
	Shader    rl.Shader
	LightType LightType
	Position  Vector3
	Target    Vector3
	Color     rl.Color
	Intensity float32
	Enabled   int32
	// shader locations
	EnabledLoc int32
	TypeLoc    int32
	PosLoc     int32
	TargetLoc  int32
	ColorLoc   int32
}

const maxLightsCount = 4

var lightCount = 0

func NewLight(
	lightType LightType,
	position, target Vector3,
	color rl.Color,
	intensity float32,
	shader rl.Shader,
) Light {
	light := Light{
		Shader: shader,
	}
	if lightCount < maxLightsCount {
		light.Enabled = 1
		light.LightType = lightType
		light.Position = position
		light.Target = target
		light.Color = color
		light.Intensity = intensity
		light.EnabledLoc = rl.GetShaderLocation(shader, fstr("lights[%d].enabled", lightCount))
		light.TypeLoc = rl.GetShaderLocation(shader, fstr("lights[%d].type", lightCount))
		light.PosLoc = rl.GetShaderLocation(shader, fstr("lights[%d].position", lightCount))
		light.TargetLoc = rl.GetShaderLocation(shader, fstr("lights[%d].target", lightCount))
		light.ColorLoc = rl.GetShaderLocation(shader, fstr("lights[%d].color", lightCount))
		light.Update()
		lightCount++
	}
	return light
}

func (self *Light) Update() {
	rl.DrawSphere(self.Position, 0.2, self.Color)

	// Send to shader light enabled state and type
	rl.SetShaderValue(self.Shader, self.EnabledLoc, unsafe.Slice((*float32)(unsafe.Pointer(&self.Enabled)), 4), rl.ShaderUniformInt)
	rl.SetShaderValue(self.Shader, self.TypeLoc, unsafe.Slice((*float32)(unsafe.Pointer(&self.LightType)), 4), rl.ShaderUniformInt)

	// Send to shader light position values
	rl.SetShaderValue(self.Shader, self.PosLoc, []float32{self.Position.X, self.Position.Y, self.Position.Z}, rl.ShaderUniformVec3)

	// Send to shader light target values
	rl.SetShaderValue(self.Shader, self.TargetLoc, []float32{self.Target.X, self.Target.Y, self.Target.Z}, rl.ShaderUniformVec3)

	// Send to shader light color values
	rl.SetShaderValue(
		self.Shader, self.ColorLoc,
		[]float32{
			float32(self.Color.R) * self.Intensity / 255,
			float32(self.Color.G) * self.Intensity / 255,
			float32(self.Color.B) * self.Intensity / 255,
			float32(self.Color.A) * self.Intensity / 255,
		},
		rl.ShaderUniformVec4,
	)
}

File: MikeAsh-C++-3D/src/Light.cpp
#include "../include/Light.hpp"

#include <raylib.h>

#include <cstdio>
#include <format>

#include "../include/Engine.hpp"

Light::Light(LightType light_type, v3 position, v3 target, Color color, float intensity,
             Shader shader)
    : shader(shader),
      light_type(light_type),
      position(position),
      target(target),
      color(color),
      intensity(intensity) {
    if (light_count < MAX_LIGHTS_COUNT) {
        enabled = 1;
        std::string prefix = std::format("lights[{}]", light_count);

        enabled_loc = GetShaderLocation(shader, (prefix + ".enabled").c_str());
        type_loc = GetShaderLocation(shader, (prefix + ".type").c_str());
        pos_loc = GetShaderLocation(shader, (prefix + ".position").c_str());
        target_loc = GetShaderLocation(shader, (prefix + ".target").c_str());
        color_loc = GetShaderLocation(shader, (prefix + ".color").c_str());
        update();
        light_count++;
    }
}

void Light::update() {
    DrawSphere(position, 0.2, color);

    // Send to shader light enabled state and type
    SetShaderValue(shader, enabled_loc, &enabled, SHADER_UNIFORM_INT);
    SetShaderValue(shader, type_loc, &light_type, SHADER_UNIFORM_INT);

    // Send to shader light position values
    SetShaderValue(shader, pos_loc, &position, SHADER_UNIFORM_VEC3);

    // Send to shader light target values
    SetShaderValue(shader, target_loc, &target, SHADER_UNIFORM_VEC3);

    // Send to shader light color values
    SetShaderValue(shader, color_loc,
                   (float[]){
                       color.r * intensity / 255,
                       color.g * intensity / 255,
                       color.b * intensity / 255,
                       color.a * intensity / 255,
                   },
                   SHADER_UNIFORM_VEC4);
}

void Light::SetPosition(v3 position) { this->position = position; }
void Light::SetTarget(v3 target) { this->target = target; }
